src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -3809,6 +3774,141 @@ Node TheoryStringsRewriter::getStringOrEmpty(Node n)
   return res;
 }
 
+bool TheoryStringsRewriter::inferZerosInSumGeq(Node x,
+                                               std::vector<Node>* ys,
+                                               std::vector<Node>* zeroYs)
+{
+  Assert(zeroYs->empty());
+
+  NodeManager* nm = NodeManager::currentNM();
+
+  // Check if we can show that y1 + ... + yn >= x
+  Node sum = (ys->size() > 1) ? nm->mkNode(kind::PLUS, *ys) : (*ys)[0];
+  if (!checkEntailArith(sum, x))
+  {
+    return false;
+  }
+
+  // Try to remove yi one-by-one and check if we can still show:
+  //
+  // y1 + ... + yi-1 +  yi+1 + ... + yn >= x
+  //
+  // If that's the case, we know that yi can be zero and the inequality still
+  // holds.
+  size_t i = 0;
+  while (i < ys->size())
+  {
+    Node yi = (*ys)[i];
+    std::vector<Node>::iterator pos = ys->erase(ys->begin() + i);
+    if (ys->size() > 1)
+    {
+      sum = nm->mkNode(kind::PLUS, *ys);
+    }
+    else
+    {
+      sum = ys->size() == 1 ? (*ys)[0] : nm->mkConst(Rational(0));
+    }
+
+    if (checkEntailArith(sum, x))
+    {
+      zeroYs->push_back(yi);
+    }
+    else
+    {
+      ys->insert(pos, yi);
+      i++;
+    }
+  }
+  return true;
+}
+
+Node TheoryStringsRewriter::inferEqsFromContains(Node x, Node y)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  Node emp = nm->mkConst(String(""));
+
+  Node xLen = nm->mkNode(kind::STRING_LENGTH, x);
+  std::vector<Node> yLens;
+  if (y.getKind() != kind::STRING_CONCAT)
+  {
+    yLens.push_back(nm->mkNode(kind::STRING_LENGTH, y));
+  }
+  else
+  {
+    for (const Node& yi : y)
+    {
+      yLens.push_back(nm->mkNode(kind::STRING_LENGTH, yi));
+    }
+  }
+
+  std::vector<Node> zeroLens;
+  if (x == emp)
+  {
+    // If x is the empty string, then all ys must be empty, too, and we can
+    // skip the expensive checks. Note that this is just a performance
+    // optimization.
+    zeroLens.swap(yLens);
+  }
+  else
+  {
+    // Check if we can infer that str.len(x) <= str.len(y). If that is the
+    // case, try to minimize the sum in str.len(x) <= str.len(y1) + ... +
+    // str.len(yn) (where y = y1 ++ ... ++ yn) while keeping the inequality
+    // true. The terms that can have length zero without making the inequality
+    // false must be all be empty if (str.contains x y) is true.
+    if (!inferZerosInSumGeq(xLen, &yLens, &zeroLens))
+    {
+      // We could not prove that the inequality holds
+      return Node::null();
+    }
+    else if (yLens.size() == y.getNumChildren())
+    {
+      // We could only prove that the inequality holds but not that any of the
+      // ys must be empty
+      return nm->mkNode(kind::EQUAL, x, y);
+    }
+  }
+
+  if (y.getKind() != kind::STRING_CONCAT)
+  {
+    if (zeroLens.size() == 1)
+    {
+      // y is not a concatenation and we found that it must be empty, so just
+      // return (= y "")
+      Assert(zeroLens[0][0] == y);
+      return nm->mkNode(kind::EQUAL, y, emp);
```

Comment:
Remove the kind:: from places (we are in the kind namespace).

Reply:
I somewhat prefer to have enums mentioned explicitly in their namespace to avoid confusion. But yeah, consistency is probably more important.